"""
Pipeline for calling and genotyping Structural Variants from long-read sequencing in a single individual
"""
import os

configfile: "config/config.yaml"

# Import sample config (sample name, sra accession)
import pandas as pd

samples = (
    pd.read_csv(config["samples"], sep="\t", dtype={"sample_name": str, "sra": str, "genome": str})
    .set_index("sample_name", drop=False)
    .sort_index()
)

# Only accepts a single genome and a single sample/individual
# genome = config['genome']
genome = samples["genome"].iloc[0]
sample_id = samples["sample_name"].iloc[0]

wdir = config['datadir'] + genome

aligner = ["minimap2","nglmr"]

wildcard_constraints:
    wdir = wdir,
    genome = genome,
    sample_id = sample_id,
    aligner = aligner

include: "common.smk"
include: "rules/common.smk"
include: "rules/data_download.smk"
include: "rules/data_qc.smk"
include: "rules/mapping.smk"
include: "rules/ensemblecalling.smk"
include: "rules/merging.smk"
include: "rules/genotyping.smk"
include: "rules/report.smk"
include: "rules/mappability.smk"


rule all:
    input:
        expand("{wdir}/{genome}_filtered.vcf.gz", wdir=wdir, genome=genome),
        expand("{wdir}/{genome}_finalQC.html", wdir=wdir, genome=genome)
