---
title: "SV calling from long reads in a single individual"
date: "`r Sys.Date()`"
output: html_document
params:
  wdir: ""
  genome: ""
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r packages, include=FALSE}
require(rmarkdown)
require(vcfR)
require(adegenet)
require(poppr)
require(jsonlite)
require(kableExtra)
require(ggplot2)
require(ggpubr)
require(stringr)
require(eulerr)
```


# Assembly


## Assembly summary statistics

```{r assemblysummary, echo = F}
json = jsonlite::read_json(paste0("../../", wdir, "/", genome, "_assembly_data_report.jsonl"))


cat("Accession:", json$accession, "\n")

cat("Bioproject Accession:", json$assemblyInfo$bioprojectAccession, "\n")

cat("Total sequence length:", json$assemblyStats$totalSequenceLength, "\n")
cat("Total ungapped length:", json$assemblyStats$totalUngappedLength, "\n")

cat("Genome coverage:", json$assemblyStats$genomeCoverage, "\n")

cat("Number of contigs:", json$assemblyStats$numberOfContigs, "\n")
cat("Contig N50 length (Mb):", json$assemblyStats$contigN50/10^6, "\n")

cat("Number of scaffolds:", json$assemblyStats$numberOfScaffolds, "\n")
cat("Scaffold N50 length (Mb):", json$assemblyStats$scaffoldN50/10^6, "\n")

cat("GC percent:", json$assemblyStats$gcPercent, "\n")
```

## Chromosomes


```{r chromosomes, echo = F}
dat = tidyjson::read_json(paste0("../../", wdir, "/", genome, "_sequence_report.jsonl"), format = "jsonl")
chr = data.frame(idx = dat$document.id)

chr$seqname = NA
chr$genbankaccession = NA
chr$type = NA
chr$length = NA
chr$gcpercent = NA

for (i in 1:nrow(chr)) {
  chr$seqname[i] = dat$..JSON[[i]]$chrName
  chr$length[i] = dat$..JSON[[i]]$length/10^6
  chr$type[i] = dat$..JSON[[i]]$assignedMoleculeLocationType
  chr$gcpercent[i] = ifelse(!is.null(dat$..JSON[[i]]$gcPercent), dat$..JSON[[i]]$gcPercent, NA)
  chr$genbankaccession[i] = dat$..JSON[[i]]$genbankAccession
}

genome_size = sum(chr$length)
```



```{r chromosometable, echo=FALSE, message=FALSE, warning=FALSE}
kable(chr[,-1], digits = 2, align = "c", caption = "Summary of chromosome assembly.", col.names = c("Name", "Genbank Accession", "Type", "Length (Mb)", "GC percent"), format = "html", table.attr = "style='width:80%;'") %>% kableExtra::kable_styling()
```


# Mapping


```{r mapping, echo = F}
mapping = readLines(paste0("../../", wdir, "/mapping/", genome, "_mapping.stats.tsv"))

print(gsub("\t", " ", mapping))
```


The diagnostic plots for the mapping with minimap2 are in the `mapping/` directory.



# SV calling

```{r loadingSV, echo=FALSE, results=FALSE, message=FALSE, warning=FALSE}
sniffles = read.vcfR(paste0("../../", wdir, "/", genome, "_sniffles_noBND.vcf"))
gt.sniffles = as.data.frame(sniffles@gt)
fix.sniffles = as.data.frame(sniffles@fix)
gt.sniffles$Caller = "sniffles"
fix.sniffles$Caller = "sniffles"

svim = read.vcfR(paste0("../../", wdir, "/", genome, "_svim_noBND.vcf"))
gt.svim = as.data.frame(svim@gt)
fix.svim = as.data.frame(svim@fix)
gt.svim$Caller = "svim"
fix.svim$Caller = "svim"

cutesv = read.vcfR(paste0("../../", wdir, "/", genome, "_cutesv_noBND.vcf"))
gt.cutesv = as.data.frame(cutesv@gt)
fix.cutesv = as.data.frame(cutesv@fix)
gt.cutesv$Caller = "cutesv"
fix.cutesv$Caller = "cutesv"

debreak = read.vcfR(paste0("../../", wdir, "/", genome, "_debreak_noBND.vcf"))
gt.debreak = as.data.frame(debreak@gt)
fix.debreak = as.data.frame(debreak@fix)
gt.debreak$Caller = "debreak"
fix.debreak$Caller = "debreak"

colnames(gt.sniffles) = c("Format", "Sample", "Caller")
colnames(gt.svim) = c("Format", "Sample", "Caller")
colnames(gt.cutesv) = c("Format", "Sample", "Caller")
colnames(gt.debreak) = c("Format", "Sample", "Caller")

gt = rbind(gt.sniffles,
           gt.svim,
           gt.cutesv,
           gt.debreak)

fix = rbind(fix.sniffles,
           fix.svim,
           fix.cutesv,
           fix.debreak)

rm(fix.sniffles,
           fix.svim,
           fix.cutesv,
           fix.debreak)

rm(gt.sniffles,
           gt.svim,
           gt.cutesv,
           gt.debreak)
gc()
```


```{r parseSV, echo = F}
fix$sv_type = as.vector(gsub("SVTYPE=", "", str_match(fix$INFO, "SVTYPE=[A-Z]+")))
fix$sv_length = as.numeric(as.vector(gsub("SVLEN=", "", str_match(fix$INFO, "SVLEN=[+-]*[0-9]+"))))
```


## Number of SV detected by each caller

```{r SVcalling, echo = T}
table(fix$Caller)
```

## SV quality score


*Note that all SV callers do not use the same way of calculating the Quality Score and they are not all on the same scale (e.g. log/phred-scale).*


```{r SVquality, echo = F, warning = F}
ggplot(fix, aes(x = Caller, y = as.numeric(QUAL), colour = Caller, group = Caller)) +
  geom_boxplot() +
  xlab("SV caller") + ylab("Quality Score") +
  theme_bw()
```


```{r SVqualitydensity, echo = F, warning = F}
ggplot(fix, aes(x = as.numeric(QUAL), colour = Caller, group = Caller)) +
  geom_density() +
  xlab("Quality Score") + ylab("Density") +
  facet_wrap(~ Caller, scales = "free") +
  theme_bw()
```



## SV type

```{r SVtypetable, echo = T}
table(fix$sv_type, fix$Caller)
```



```{r SVtype, echo = F, warning = F}
ggplot(fix, aes(x = sv_type, fill = sv_type, colour = sv_type)) +
  geom_bar() +
  xlab("SV type") + ylab("Count") +
  facet_wrap(~ Caller, scales = "fixed") +
  theme_bw()
```


## SV length


```{r SVlength, echo = F, warning = F}
df = aggregate(sv_length ~ sv_type + Caller, fix, function(x) {sum(abs(x))})

ggplot(df, aes(x = sv_type, y = sv_length/10^6, fill = sv_type, colour = sv_type)) +
  geom_bar(stat = "identity") +
  xlab("SV type") + ylab("Total absolute length (Mb)") +
  facet_wrap(~ Caller, scales = "fixed") +
  theme_bw()
```




```{r SVproportion, echo = F, warning = F}
ggplot(df, aes(x = sv_type, y = sv_length/(genome_size * 10^6), fill = sv_type, colour = sv_type)) +
  geom_bar(stat = "identity") +
  xlab("SV type") + ylab("Proportion of the genome") +
  facet_wrap(~ Caller, scales = "fixed") +
  ylim(0, 1) +
  theme_bw()
```




# Merging with Jasmine

```{r loadingSV2, echo=FALSE, results=FALSE, message=FALSE, warning=FALSE}
merged = read.vcfR(paste0("../../", wdir, "/", genome, "_merged.vcf"))
gt.merged = as.data.frame(merged@gt)
fix.merged = as.data.frame(merged@fix)

fix.merged$sv_type = as.vector(gsub("SVTYPE=", "", str_match(fix.merged$INFO, "SVTYPE=[A-Z]+")))
fix.merged$sv_length = as.numeric(as.vector(gsub("SVLEN=", "", str_match(fix.merged$INFO, "SVLEN=[-]*[0-9]+"))))

fix.merged$supp_vec = as.vector(gsub("SUPP_VEC=", "", str_match(fix.merged$INFO, "SUPP_VEC=[0-1]+")))
fix.merged$n_supp = as.numeric(as.vector(gsub("SUPP=", "", str_match(fix.merged$INFO, "SUPP=[0-1]+"))))


fix.merged$svim = ifelse(lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 1) == "1", 1, 0)
fix.merged$sniffles = ifelse(lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 2) == "1", 1, 0)
fix.merged$cutesv = ifelse(lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 3) == "1", 1, 0)
fix.merged$debreak = ifelse(lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 4) == "1", 1, 0)
```

<!-- 
## Number of SV for each tool before/after merging (and per SV type)


**TODO**

Number of SV supported by at least two tools:



Number of SV supported by at least three tools:


Number of SV supported by the four tools: -->


## Venn diagram - All SV types together


```{r venndiag, echo = F, fig.width= 12, fig.height=9}
plot(euler(fix.merged[, colnames(fix.merged) %in% c("svim", "cutesv", "sniffles", "debreak")], shape = "ellipse"),
     quantities = TRUE,
     fills = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3), alpha('#ff4500',0.3)))
```



## Venn diagram - DEL



```{r venndiagdel, echo = F, fig.width= 12, fig.height=9}
fix.merged.del = fix.merged[which(fix.merged$sv_type == "DEL"),]
plot(euler(fix.merged.del[, colnames(fix.merged.del) %in% c("svim", "cutesv", "sniffles", "debreak")], shape = "ellipse"),
     quantities = TRUE,
     fills = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3), alpha('#ff4500',0.3)))
```


## Venn diagram - INS



```{r venndiagins, echo = F, fig.width= 12, fig.height=9}
fix.merged.ins = fix.merged[which(fix.merged$sv_type == "INS"),]
plot(euler(fix.merged.ins[, colnames(fix.merged.ins) %in% c("svim", "cutesv", "sniffles", "debreak")], shape = "ellipse"),
     quantities = TRUE,
     fills = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3), alpha('#ff4500',0.3)))
```

## Venn diagram - DUP


```{r venndiagdup, echo = F, fig.width= 12, fig.height=9}
fix.merged.dup = fix.merged[which(fix.merged$sv_type == "DUP"),]
plot(euler(fix.merged.dup[, colnames(fix.merged.dup) %in% c("svim", "cutesv", "sniffles", "debreak")], shape = "ellipse"),
     quantities = TRUE,
     fills = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3), alpha('#ff4500',0.3)))
```


## Venn diagram - INV


```{r venndiaginv, echo = F, fig.width= 12, fig.height=9}
fix.merged.inv = fix.merged[which(fix.merged$sv_type == "INV"),]
plot(euler(fix.merged.inv[, colnames(fix.merged.inv) %in% c("svim", "cutesv", "sniffles", "debreak")], shape = "ellipse"),
     quantities = TRUE,
     fills = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3), alpha('#ff4500',0.3)))
```





# Genotyping and final callset


```{r loadingFinal, echo = F, results=FALSE}
geno = read.vcfR(paste0("../../", wdir, "/", genome, "_filtered.vcf"))
gt.geno = as.data.frame(geno@gt)
fix.geno = as.data.frame(geno@fix)
```


```{r samplename, echo = F}
colnames(gt.geno)[2]
```



```{r genotyping, echo = F}
geno
```


## SV type and length


```{r formatGeno, echo = F}
fix.geno$sv_type = as.vector(gsub("SVTYPE=", "", str_match(fix.geno$INFO, "SVTYPE=[A-Z]+")))
fix.geno$sv_length = as.numeric(as.vector(gsub("SVLEN=", "", str_match(fix.geno$INFO, "SVLEN=[+-]*[0-9]+"))))

fix.geno$sv_support = as.numeric(as.vector(gsub("SUPPORT=", "", str_match(fix.geno$INFO, "SUPPORT=[+-]*[0-9]+"))))
```



```{r SVtablefinal}
table(fix.geno$sv_type)
```


The stacked barplot of SV count for each SV type and SV length (no filter on quality scores and quality score > 10).

```{r SVtypefinal, echo = F, warning = F}
p1 = ggplot(fix.geno, aes(x = abs(sv_length), fill = sv_type, colour = sv_type)) +
  geom_bar() +
  xlab("Length of Structural Variant (bp)") + ylab("SV Count") +
  xlim(0, 1000) +
  theme_bw()

p2 = ggplot(fix.geno, aes(x = abs(sv_length), fill = sv_type, colour = sv_type)) +
  geom_bar() +
  xlab("Length of Structural Variant (bp)") + ylab("SV Count") +
  xlim(0, 50000) +
  theme_bw()

ggarrange(p1, p2, nrow = 2)
```



After filtering at QC > 10.


```{r SVtablefinal2}
table(fix.geno$sv_type[which(fix.geno$QUAL > 10)])
```



```{r SVtypefinal2, echo = F, warning = F}
p1 = ggplot(fix.geno[which(fix.geno$QUAL > 10),], aes(x = abs(sv_length), fill = sv_type, colour = sv_type)) +
  geom_bar() +
  xlab("Length of Structural Variant (bp)") + ylab("SV Count") +
  xlim(0, 1000) +
  theme_bw()

p2 = ggplot(fix.geno[which(fix.geno$QUAL > 10),], aes(x = abs(sv_length), fill = sv_type, colour = sv_type)) +
  geom_bar() +
  xlab("Length of Structural Variant (bp)") + ylab("SV Count") +
  xlim(0, 50000) +
  theme_bw()

ggarrange(p1, p2, nrow = 2)
```




```{r SVlengthfinal, echo = F, warning = F}
fix.geno$sv_type = as.factor(fix.geno$sv_type)

p1 = ggplot(fix.geno, aes(y = abs(sv_length), x = sv_type, fill = sv_type, colour = sv_type)) +
  geom_boxplot() +
  ylab("SV Length (bp)") +
  theme_bw()


p2 = ggplot(fix.geno, aes(y = log10(abs(sv_length)), x = sv_type, fill = sv_type, colour = sv_type)) +
  geom_boxplot() +
  ylab("SV Length (log10 scale)") +
  theme_bw()

ggarrange(p1, p2, nrow = 2)
```



## Missing data


```{r parsinggeno, echo=FALSE, message=FALSE, warning=FALSE}
gt.geno$GT = unlist(lapply(strsplit(gt.geno[,2], ":"), `[[`, 1))
gt.geno$AD = as.numeric(unlist(lapply(strsplit(gt.geno[,2], ":"), `[[`, 2)))
gt.geno$DP = unlist(lapply(strsplit(gt.geno[,2], ":"), `[[`, 3))
gt.geno$DP_REF = as.numeric(unlist(lapply(strsplit(gt.geno$DP, ","), `[[`, 1)))
gt.geno$DP_ALT = as.numeric(unlist(lapply(strsplit(gt.geno$DP, ","), `[[`, 2)))
gt.geno$PL = unlist(lapply(strsplit(gt.geno[,2], ":"), `[[`, 4))
gt.geno$PL_HOMREF = as.numeric(unlist(lapply(strsplit(gt.geno$PL, ","), `[[`, 1)))
gt.geno$PL_HET = as.numeric(unlist(lapply(strsplit(gt.geno$PL, ","), `[[`, 2)))
gt.geno$PL_HOMALT = as.numeric(unlist(lapply(strsplit(gt.geno$PL, ","), `[[`, 3)))
```


```{r missinggeno}
cat("Proportion of missing data (./.):", sum(gt.geno$GT == "./.")/nrow(gt.geno), "\n")

cat("Proportion of partially missing data (./?):", sum(gt.geno$GT %in% c("./0", "0/.", "1/.", "./1"))/nrow(gt.geno), "(should be 0 if everything is ok) \n")
```

## Heterozygosity


```{r HeGeno}
cat("Proportion of heterozygous SV (0/1 or 1/0):", sum(gt.geno$GT %in% c("1/0", "0/1"))/nrow(gt.geno), "\n")
```


We investigated the usual Variant Calling quality criterion:

* DP: Approximate read depth; some reads may have been filtered
* AD: Allelic depths for the ref and alt alleles in the order listed
* PL: Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification (lower score is better, lower score = most likely genotype)


## Supporting reads (DP:AD)



Approximate read depth (should be balanced between reference and variant if unbiased).


```{r SV_DP, echo = F, warning = F}
ggplot(gt.geno, aes(x = DP_REF, y = DP_ALT, group = GT, colour = GT)) +
  geom_point(alpha = 0.3) +
  xlab("Reference allele read depth (DP)") + ylab("Variant allele read depth (DP)") +
  theme_bw()
```


Variant allelic depth (higher for heterozygotes is better).


```{r AD_GT, echo = F, warning = F}
gt.geno$GT = as.factor(gt.geno$GT)

p1 = ggplot(gt.geno, aes(x = GT, y = AD, group = GT, colour = GT)) +
  geom_boxplot() +
  ylab("Variant allelic depth (AD)") +
  theme_bw()

p2 = ggplot(gt.geno, aes(x = GT, y = AD, group = GT, colour = GT)) +
  geom_boxplot() +
  ylab("Variant allelic depth (AD)") +
  ylim(0, 100) +
  theme_bw()

ggarrange(p1, p2, ncol = 2)
```


## Phred-scale likelihoods for heterozygotes

Lower phred-scale likelihood is better.



Scores for 0/0 (homozygous ref), 0/1 (heterozygous), and 1/1 (homozygous alt) genotypes. For a phred-scaled likelihood of P, the raw likelihood of that genotype L = 10-P/10 (so the higher the number, the less likely it is that your sample is that genotype). The sum of likelihoods is not necessarily 1. 



```{r PL, echo = F, warning = F}
df = data.frame(GT = c(rep("0/0", sum(gt.geno$GT %in% c("0/0"))),
                       rep("0/1", sum(gt.geno$GT %in% c("0/1"))),
                       rep("1/1", sum(gt.geno$GT %in% c("1/1")))),
                PL = c(gt.geno$PL_HOMREF[which(gt.geno$GT %in% c("0/0"))],
                       gt.geno$PL_HET[which(gt.geno$GT %in% c("0/1"))],
                       gt.geno$PL_HOMALT[which(gt.geno$GT %in% c("1/1"))]))
df$GT = as.factor(df$GT)


p1 = ggplot(df, aes(x = GT, y = PL, group = GT, colour = GT)) +
  geom_boxplot() +
  ylab("Phred-scaled genotype likelihood") +
  theme_bw()

p2 = ggplot(df, aes(x = GT, y = PL, group = GT, colour = GT)) +
  geom_boxplot() +
  ylab("Phred-scaled genotype likelihood") +
  ylim(0, 100) +
  theme_bw()

ggarrange(p1, p2, ncol = 2)
```




# SV and genomic landscapes



## Quality and support along the genome

Quality scores and allele support are expected to homogeneous along the genome.


Quality Score.

```{r QClandscape, fig.height = 24, echo = F, warning = F}
ggplot(fix.geno, aes(x = as.numeric(POS)/10^6, y = as.numeric(QUAL))) +
  geom_point(alpha = 0.1) +
  geom_smooth(method = "gam", se = T) +
  xlab("Position (Mb)") + ylab("SV Quality score") +
  facet_wrap(~ CHROM, scales = "free_y", ncol = 3) +
  theme_bw()
```


Support for variant allele.

```{r suppportlandscape, fig.height = 24, echo = F, warning = F}
ggplot(fix.geno, aes(x = as.numeric(POS)/10^6, y = as.numeric(sv_support))) +
  geom_point(alpha = 0.1) +
  geom_smooth(method = "gam", se = T) +
  xlab("Position (Mb)") + ylab("SV support for allele variant") +
  facet_wrap(~ CHROM, scales = "free_y", ncol = 3) +
  theme_bw()
```






## The distribution of SVs along the genome


```{r SVlandscape, fig.height = 24, echo = F, warning = F}
ggplot(fix.geno, aes(x = as.numeric(POS)/10^6)) +
  geom_histogram(bins = 1000) +
  xlab("Position (Mb)") + ylab("SV count") +
  facet_wrap(~ CHROM, ncol = 3)
```




