---
title: "SV calling from long reads in a single individual"
date: "`r Sys.Date()`"
output: html_document
params:
  wdir: ""
  genome: ""
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r packages, include=FALSE}
require(rmarkdown)
require(vcfR)
require(adegenet)
require(poppr)
require(jsonlite)
require(kableExtra)
require(ggplot2)
require(ggpubr)
require(stringr)
require(eulerr)
```


# Assembly


## Assembly summary statistics

```{r assemblysummary, echo = F}
json = jsonlite::read_json(paste0("../../", wdir, "/", genome, "_assembly_data_report.jsonl"))


cat("Accession:", json$accession, "\n")

cat("Bioproject Accession:", json$assemblyInfo$bioprojectAccession, "\n")

cat("Total sequence length:", json$assemblyStats$totalSequenceLength, "\n")
cat("Total ungapped length:", json$assemblyStats$totalUngappedLength, "\n")

cat("Genome coverage:", json$assemblyStats$genomeCoverage, "\n")

cat("Number of contigs:", json$assemblyStats$numberOfContigs, "\n")
cat("Contig N50 length (Mb):", json$assemblyStats$contigN50/10^6, "\n")

cat("Number of scaffolds:", json$assemblyStats$numberOfScaffolds, "\n")
cat("Scaffold N50 length (Mb):", json$assemblyStats$scaffoldN50/10^6, "\n")

cat("GC percent:", json$assemblyStats$gcPercent, "\n")
```

## Chromosomes


```{r chromosomes, echo = F}
dat = tidyjson::read_json(paste0("../../", wdir, "/", genome, "_sequence_report.jsonl"), format = "jsonl")
chr = data.frame(idx = dat$document.id)

chr$seqname = NA
chr$genbankaccession = NA
chr$type = NA
chr$length = NA
chr$gcpercent = NA

for (i in 1:nrow(chr)) {
  chr$seqname[i] = dat$..JSON[[i]]$chrName
  chr$length[i] = dat$..JSON[[i]]$length/10^6
  chr$type[i] = dat$..JSON[[i]]$assignedMoleculeLocationType
  chr$gcpercent[i] = ifelse(!is.null(dat$..JSON[[i]]$gcPercent), dat$..JSON[[i]]$gcPercent, NA)
  chr$genbankaccession[i] = dat$..JSON[[i]]$genbankAccession
}

genome_size = sum(chr$length)
```



```{r chromosometable, echo=FALSE, message=FALSE, warning=FALSE}
kable(chr[,-1], digits = 2, align = "c", caption = "Summary of chromosome assembly.", col.names = c("Name", "Genbank Accession", "Type", "Length (Mb)", "GC percent"), format = "html", table.attr = "style='width:80%;'") %>% kableExtra::kable_styling()
```


# Mapping


```{r mapping_minimap2, echo = F}
mapping_minimap2 = readLines(paste0("../../", wdir, "/mapping/", genome, "_minimap2_mapping.stats.tsv"))

print("Mapping with minimap2")
print(gsub("\t", " ", mapping_minimap2))
```




```{r mapping_ngmlr, echo = F}
mapping_ngmlr = readLines(paste0("../../", wdir, "/mapping/", genome, "_ngmlr_mapping.stats.tsv"))

print("Mapping with ngmlr")
print(gsub("\t", " ", mapping_ngmlr))
```




The diagnostic plots for the mapping with minimap2 and ngmlr are in the `mapping/` directory.



# SV calling


The final VCF produced contains 9 samples:


* eight samples are individual callset from a pair of aligner + sv caller
* the last sample is the final genotype call after merging callsets with Jasmine


Below are presented the callsets for the different tools (aligner  + SV caller) before the merging with Jasmine. Hence some calls may not be represented in the final sample (lost during merging).


How to make the difference between a call that has been merged and another that has not (called by an individual tool but removed during merging)?

"SUPP_VEC=" is not present in the INFO field.


```{r loading_final_vcf, echo=FALSE, results=FALSE, message=FALSE, warning=FALSE}
# merged = read.vcfR(paste0("../../", wdir, "/", genome, "_filtered.vcf"))
merged = read.table(paste0("../../", wdir, "/", genome, "_filtered.tsv"),
                    header = TRUE, sep = "\t")

# TESTING
# merged = read.vcfR(paste0("../../", wdir, "/", "test_lightheader.vcf"), check_keys = FALSE,
#                    verbose = TRUE)
# merged = read.table(paste0("../../", wdir, "/", "test_lightheader.tsv"),
#                     header = TRUE, sep = "\t", quote = '"')

# saveRDS(merged, "../../data/test.Rds")
# merged = readRDS("../../data/test.Rds")
# 
# gt.merged = as.data.frame(merged@gt)
# fix.merged = as.data.frame(merged@fix)
fix.merged = merged[,1:6]
gt.merged = merged[,7:16]

fix.merged$sv_type = as.vector(gsub("SVTYPE=", "", str_match(fix.merged$INFO, "SVTYPE=[A-Z]+")))
fix.merged$sv_length = as.numeric(as.vector(gsub("SVLEN=", "", str_match(fix.merged$INFO, "SVLEN=[-]*[0-9]+"))))

fix.merged$supp_vec = as.vector(gsub("SUPP_VEC=", "", str_match(fix.merged$INFO, "SUPP_VEC=[0-1]+")))
fix.merged$n_supp = as.numeric(as.vector(gsub("SUPP=", "", str_match(fix.merged$INFO, "SUPP=[0-9]+"))))
fix.merged$jasmine = ifelse(!is.na(fix.merged$supp_vec), 1, 0)
```

```{r samplenames, echo = T}
# sample_names = c("FORMAT", paste(c("minimap2_sniffles", "minimap2_svim", "minimap2_cutesv",
#                                    "minimap2_debreak", "ngmlr_sniffles", "ngmlr_svim",
#                                    "ngmlr_cutesv", "ngmlr_debreak"), genome, sep = "_"), genome)
# names(gt.merged) = sample_names
```


```{r get_tool_support, echo = FALSE}
# Get aligner support
fix.merged$minimap2 = ifelse(lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 1) == "1" |
                               lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 2) == "1" |
                               lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 3) == "1" |
                               lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 4) == "1", 1, 0)
fix.merged$ngmlr = ifelse(lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 5) == "1" |
                               lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 6) == "1" |
                               lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 7) == "1" |
                               lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 8) == "1", 1, 0)

# Get SV caller support
fix.merged$sniffles = ifelse(lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 1) == "1" | lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 5) == "1", 1, 0)
fix.merged$svim = ifelse(lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 2) == "1" | lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 6) == "1", 1, 0)
fix.merged$cutesv = ifelse(lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 3) == "1" | lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 7) == "1", 1, 0)
fix.merged$debreak = ifelse(lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 4) == "1" | lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 8) == "1", 1, 0)
```





```{r loadingSV, echo=FALSE, results=FALSE, message=FALSE, warning=FALSE}
# sniffles = read.vcfR(paste0("../../", wdir, "/", genome, "_sniffles_noBND.vcf"))
# gt.sniffles = as.data.frame(sniffles@gt)
# fix.sniffles = as.data.frame(sniffles@fix)
# gt.sniffles$Caller = "sniffles"
# fix.sniffles$Caller = "sniffles"
# 
# svim = read.vcfR(paste0("../../", wdir, "/", genome, "_svim_noBND.vcf"))
# gt.svim = as.data.frame(svim@gt)
# fix.svim = as.data.frame(svim@fix)
# gt.svim$Caller = "svim"
# fix.svim$Caller = "svim"
# 
# cutesv = read.vcfR(paste0("../../", wdir, "/", genome, "_cutesv_noBND.vcf"))
# gt.cutesv = as.data.frame(cutesv@gt)
# fix.cutesv = as.data.frame(cutesv@fix)
# gt.cutesv$Caller = "cutesv"
# fix.cutesv$Caller = "cutesv"
# 
# debreak = read.vcfR(paste0("../../", wdir, "/", genome, "_debreak_noBND.vcf"))
# gt.debreak = as.data.frame(debreak@gt)
# fix.debreak = as.data.frame(debreak@fix)
# gt.debreak$Caller = "debreak"
# fix.debreak$Caller = "debreak"
# 
# colnames(gt.sniffles) = c("Format", "Sample", "Caller")
# colnames(gt.svim) = c("Format", "Sample", "Caller")
# colnames(gt.cutesv) = c("Format", "Sample", "Caller")
# colnames(gt.debreak) = c("Format", "Sample", "Caller")
# 
# gt = rbind(gt.sniffles,
#            gt.svim,
#            gt.cutesv,
#            gt.debreak)
# 
# fix = rbind(fix.sniffles,
#            fix.svim,
#            fix.cutesv,
#            fix.debreak)
# 
# rm(fix.sniffles,
#            fix.svim,
#            fix.cutesv,
#            fix.debreak)
# 
# rm(gt.sniffles,
#            gt.svim,
#            gt.cutesv,
#            gt.debreak)
# gc()
```


```{r parseSV, echo = F}
fix.merged$sv_type = as.vector(gsub("SVTYPE=", "", str_match(fix.merged$INFO, "SVTYPE=[A-Z]+")))
fix.merged$sv_length = as.numeric(as.vector(gsub("SVLEN=", "", str_match(fix.merged$INFO, "SVLEN=[+-]*[0-9]+"))))
```


```{r, echo = F}
# Tidy data frame
fix.merged1 = fix.merged[fix.merged$minimap2 == 1 & fix.merged$sniffles == 1,]
fix.merged1$aligner = "minimap2"
fix.merged1$caller = "sniffles"

fix.merged2 = fix.merged[fix.merged$minimap2 == 1 & fix.merged$svim == 1,]
fix.merged2$aligner = "minimap2"
fix.merged2$caller = "svim"

fix.merged3 = fix.merged[fix.merged$minimap2 == 1 & fix.merged$cutesv == 1,]
fix.merged3$aligner = "minimap2"
fix.merged3$caller = "cutesv"

fix.merged4 = fix.merged[fix.merged$minimap2 == 1 & fix.merged$debreak == 1,]
fix.merged4$aligner = "minimap2"
fix.merged4$caller = "debreak"

fix.merged5 = fix.merged[fix.merged$ngmlr == 1 & fix.merged$sniffles == 1,]
fix.merged5$aligner = "ngmlr"
fix.merged5$caller = "sniffles"

fix.merged6 = fix.merged[fix.merged$ngmlr == 1 & fix.merged$svim == 1,]
fix.merged6$aligner = "ngmlr"
fix.merged6$caller = "svim"

fix.merged7 = fix.merged[fix.merged$ngmlr == 1 & fix.merged$cutesv == 1,]
fix.merged7$aligner = "ngmlr"
fix.merged7$caller = "cutesv"

fix.merged8 = fix.merged[fix.merged$ngmlr == 1 & fix.merged$debreak == 1,]
fix.merged8$aligner = "ngmlr"
fix.merged8$caller = "debreak"

tidy_fix = rbind(fix.merged1,fix.merged2,
                 fix.merged3,fix.merged4,
                 fix.merged5,fix.merged6,
                 fix.merged7,fix.merged8)
```

## Number of SV detected by each caller

```{r SVcalling, echo = T}
table(tidy_fix$aligner)
table(tidy_fix$caller)

table(tidy_fix$aligner, tidy_fix$caller)
```



## SV quality score


*Note that all SV callers do not use the same way of calculating the Quality Score and they are not all on the same scale (e.g. log/phred-scale).* Consequently, the QUAL score plotted here is the SVjedi-graph genotyping QUAL score, estimated in a consistent way across SV callers.

```{r SVqualityAligner, echo = F, warning = F, fig.cap = "SVjedi-graph quality score for each aligner."}
ggplot(tidy_fix, aes(x = aligner, y = as.numeric(QUAL), colour = aligner, group = aligner)) +
  geom_violin(aes(fill = aligner), color = "black") +
  geom_boxplot(alpha = 0.5, color = "black") +
  geom_point(alpha = 0.1) +
  xlab("Aligner") + ylab("Quality Score") +
  theme_bw()
```



```{r SVqualityCaller, echo = F, warning = F, fig.cap = "SVjedi-graph quality score for each aligner and SV caller."}
ggplot(tidy_fix, aes(x = caller, y = as.numeric(QUAL), colour = caller, group = caller)) +
  geom_violin(aes(fill = caller), color = "black") +
  geom_boxplot(alpha = 0.5, color = "black") +
  geom_point(alpha = 0.1) +  xlab("SV caller") + ylab("Quality Score") +
  facet_wrap(~ aligner, scales = "fixed", nrow = 2) +
  theme_bw()
```


```{r SVqualitydensity, echo = F, warning = F, fig.cap = "Density distribution of SVjedi-graph quality scores for each aligner and SV caller."}
ggplot(tidy_fix, aes(x = as.numeric(QUAL), colour = caller, group = caller)) +
  geom_density() +
  xlab("Quality Score") + ylab("Density") +
  facet_wrap(~ aligner + caller, scales = "free", ncol = 4) +
  theme_bw()
```



## SV type

```{r SVtypetable, echo = T}
table(tidy_fix$sv_type, tidy_fix$caller)
```



```{r SVtype, echo = F, warning = F, fig.cap = "Number of SVs of each type for each aligner and SV caller."}
ggplot(tidy_fix, aes(x = sv_type, fill = sv_type, colour = sv_type)) +
  geom_bar() +
  xlab("SV type") + ylab("Count") + labs(fill = "SV type", colour = "SV type") +
  facet_wrap(~ aligner + caller, scales = "fixed", ncol = 4) +
  theme_bw()
```


## SV length


```{r SVlength, echo = F, warning = F, fig.cap="Total absolute length covered by SVs. The black line is the total genome size."}
df = aggregate(sv_length ~ sv_type + caller + aligner, tidy_fix, function(x) {sum(abs(x))})

df_sample_size = aggregate(sv_length ~ sv_type + caller + aligner, tidy_fix, function(x) {length(x)})
df$sample_size = df_sample_size$sv_length
  
df$proportion_genome = df$sv_length/(genome_size * 10^6)

ggplot(df, aes(x = sv_type, y = sv_length/10^6, fill = sv_type, colour = sv_type)) +
  geom_bar(stat = "identity") +
  geom_hline(yintercept = genome_size) +
  xlab("SV type") + ylab("Total absolute length (Mb)") +
  labs(fill = "SV type", colour = "SV type") +
  facet_wrap(~ aligner + caller, scales = "fixed", ncol = 4) +
  theme_bw()
```




```{r SVproportion, echo = F, warning = F, fig.cap="Proportion of the genome covered by SVs"}
ggplot(df, aes(x = sv_type, y = proportion_genome, fill = sv_type, colour = sv_type)) +
  geom_bar(stat = "identity") +
  xlab("SV type") + ylab("Proportion of the genome") +
  facet_wrap(~ aligner + caller, scales = "fixed", ncol = 4) +
  labs(fill = "SV type", colour = "SV type") +
  # ylim(0, 1) +
  theme_bw()
```




# Merging with Jasmine

```{r loadingSV2, echo=FALSE, results=FALSE, message=FALSE, warning=FALSE}
# merged = read.vcfR(paste0("../../", wdir, "/", genome, "_merged.vcf"))
# gt.merged = as.data.frame(merged@gt)
# fix.merged = as.data.frame(merged@fix)
# 
# fix.merged$sv_type = as.vector(gsub("SVTYPE=", "", str_match(fix.merged$INFO, "SVTYPE=[A-Z]+")))
# fix.merged$sv_length = as.numeric(as.vector(gsub("SVLEN=", "", str_match(fix.merged$INFO, "SVLEN=[-]*[0-9]+"))))
# 
# fix.merged$supp_vec = as.vector(gsub("SUPP_VEC=", "", str_match(fix.merged$INFO, "SUPP_VEC=[0-1]+")))
# fix.merged$n_supp = as.numeric(as.vector(gsub("SUPP=", "", str_match(fix.merged$INFO, "SUPP=[0-1]+"))))
# 
# 
# fix.merged$svim = ifelse(lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 1) == "1", 1, 0)
# fix.merged$sniffles = ifelse(lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 2) == "1", 1, 0)
# fix.merged$cutesv = ifelse(lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 3) == "1", 1, 0)
# fix.merged$debreak = ifelse(lapply(strsplit(fix.merged$supp_vec, ""), `[[`, 4) == "1", 1, 0)
```

How many calls were merged with jasmine?


```{r, echo = TRUE}
# Absolute number
sum(fix.merged$jasmine)

# Proportion
sum(fix.merged$jasmine) / nrow(fix.merged)
```
<!-- 
## Number of SV for each tool before/after merging (and per SV type)


**TODO**

Number of SV supported by at least two tools:



Number of SV supported by at least three tools:


Number of SV supported by the four tools: -->


## Venn diagram - All SV types together

```{r venndiag1, echo = F, fig.width= 12, fig.height=9}
plot(euler(fix.merged[, colnames(fix.merged) %in% c("minimap2", "ngmlr")], shape = "ellipse"),
     quantities = TRUE,
     fills = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3), alpha('#ff4500',0.3)))
```



```{r venndiag2, echo = F, fig.width= 12, fig.height=9}
plot(euler(fix.merged[, colnames(fix.merged) %in% c("svim", "cutesv", "sniffles", "debreak")], shape = "ellipse"),
     quantities = TRUE,
     fills = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3), alpha('#ff4500',0.3)))
```



## Venn diagram - DEL



```{r venndiagdel, echo = F, fig.width= 12, fig.height=9}
fix.merged.del = fix.merged[which(fix.merged$sv_type == "DEL"),]
plot(euler(fix.merged.del[, colnames(fix.merged.del) %in% c("svim", "cutesv", "sniffles", "debreak")], shape = "ellipse"),
     quantities = TRUE,
     fills = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3), alpha('#ff4500',0.3)))
```


## Venn diagram - INS



```{r venndiagins, echo = F, fig.width= 12, fig.height=9}
fix.merged.ins = fix.merged[which(fix.merged$sv_type == "INS"),]
plot(euler(fix.merged.ins[, colnames(fix.merged.ins) %in% c("svim", "cutesv", "sniffles", "debreak")], shape = "ellipse"),
     quantities = TRUE,
     fills = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3), alpha('#ff4500',0.3)))
```

## Venn diagram - DUP


```{r venndiagdup, echo = F, fig.width= 12, fig.height=9}
fix.merged.dup = fix.merged[which(fix.merged$sv_type == "DUP"),]
plot(euler(fix.merged.dup[, colnames(fix.merged.dup) %in% c("svim", "cutesv", "sniffles", "debreak")], shape = "ellipse"),
     quantities = TRUE,
     fills = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3), alpha('#ff4500',0.3)))
```


## Venn diagram - INV


```{r venndiaginv, echo = F, fig.width= 12, fig.height=9}
fix.merged.inv = fix.merged[which(fix.merged$sv_type == "INV"),]
plot(euler(fix.merged.inv[, colnames(fix.merged.inv) %in% c("svim", "cutesv", "sniffles", "debreak")], shape = "ellipse"),
     quantities = TRUE,
     fills = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3), alpha('#ff4500',0.3)))
```




## Accumulation curve


```{r, echo = F}
#  Cumulative number of calls supported by n, n-1, ..., 1 tool
summary(fix.merged$n_supp)
table(fix.merged$n_supp)

cumsum(table(fix.merged$n_supp)[1:8])

df = data.frame(num_tools = 1:8,
                cumsum_num_sv = cumsum(table(fix.merged$n_supp)[1:8]))
```




```{r, echo = F, warning = F, fig.cap = "The cumulative number of calls (pooling tools) as a function of the number of supporting tools."}
ggplot(df, aes(x = num_tools, y = cumsum_num_sv)) +
  geom_point() +
  geom_line() +
  xlab("Minimum number of supporting tools") + ylab("Cumulative number of calls (tools pooled)") +
  ylim(0, max(df$cumsum_num_sv)) +
  theme_bw()
```






# Genotyping and final callset

Now the final callset in the merged sample is presented.


```{r loadingFinal, echo = F, results=FALSE}
# geno = read.vcfR(paste0("../../", wdir, "/", genome, "_filtered.vcf"))
# gt.geno = as.data.frame(geno@gt)
# fix.geno = as.data.frame(geno@fix)
```


```{r samplename, echo = F}
colnames(gt.merged)[-1]
```



```{r genotyping, echo = F}
# Only the final callset, all tools merged
gt.geno = gt.merged[,c(1, 10)]
```


## SV type and length


```{r SVtablefinal}
table(fix.merged$sv_type)
```


The stacked barplot of SV count for each SV type and SV length (no filter on quality scores and quality score > 10).

```{r SVtypefinal, echo = F, warning = F, fig.cap = "Joint distribution of SV length and number of SV, per SV type."}
p1 = ggplot(fix.merged, aes(x = abs(sv_length), fill = sv_type, colour = sv_type)) +
  geom_bar() +
  xlab("Length of Structural Variant (bp), less than 1 kb") + ylab("SV Count") +
  xlim(0, 1000) +
  labs(fill = "SV type", colour = "SV type") +
  theme_bw()

p2 = ggplot(fix.merged, aes(x = abs(sv_length), fill = sv_type, colour = sv_type)) +
  geom_bar() +
  xlab("Length of Structural Variant (bp), less than 50 kb") + ylab("SV Count") +
  xlim(0, 50000) +
  labs(fill = "SV type", colour = "SV type") +
  theme_bw()


p3 = ggplot(fix.merged, aes(x = abs(sv_length), fill = sv_type, colour = sv_type)) +
  geom_bar() +
  xlab("Length of Structural Variant (bp), unlimited scale") + ylab("SV Count") +
  labs(fill = "SV type", colour = "SV type") +
  theme_bw()



ggarrange(p1, p2, p3, nrow = 3)
```



After filtering at QC > 10.


```{r SVtablefinal2}
table(fix.merged$sv_type[which(fix.merged$QUAL > 10)])

# Proportion
table(fix.merged$sv_type[which(fix.merged$QUAL > 10)]) / table(fix.merged$sv_type[fix.merged$sv_type != "TRA"])
```



```{r SVtypefinal2, echo = F, warning = F, fig.cap = "Joint distribution of SV length and number of SV, per SV type (after filtering for QUAL > 10)."}
p1 = ggplot(fix.merged[which(fix.merged$QUAL > 10),], aes(x = abs(sv_length), fill = sv_type, colour = sv_type)) +
  geom_bar() +
  xlab("Length of Structural Variant (bp), less than 1 kb") + ylab("SV Count") +
  xlim(0, 1000) +
  labs(fill = "SV type", colour = "SV type") +
  theme_bw()

p2 = ggplot(fix.merged[which(fix.merged$QUAL > 10),], aes(x = abs(sv_length), fill = sv_type, colour = sv_type)) +
  geom_bar() +
  xlab("Length of Structural Variant (bp), less than 50 kb") + ylab("SV Count") +
  xlim(0, 50000) +
  labs(fill = "SV type", colour = "SV type") +
  theme_bw()

p3 = ggplot(fix.merged[which(fix.merged$QUAL > 10),], aes(x = abs(sv_length), fill = sv_type, colour = sv_type)) +
  geom_bar() +
  xlab("Length of Structural Variant (bp), unlimited scale") + ylab("SV Count") +
  labs(fill = "SV type", colour = "SV type") +
  theme_bw()

ggarrange(p1, p2, p3, nrow = 3)
```




```{r SVlengthfinal, echo = F, warning = F, fig.cap = "SV length as a function of SV type."}
fix.merged$sv_type = as.factor(fix.merged$sv_type)

p1 = ggplot(fix.merged, aes(y = abs(sv_length) / 1000, x = sv_type, fill = sv_type)) +
  geom_boxplot() +
  ylab("SV Length (kb)") + xlab("SV type") +
  labs(fill = "SV type", colour = "SV type") +
  theme_bw()


p2 = ggplot(fix.merged, aes(y = log10(abs(sv_length)), x = sv_type, fill = sv_type)) +
  geom_boxplot() +
  ylab("SV Length (log10 scale)") + xlab("SV type") +
  labs(fill = "SV type", colour = "SV type") +
  theme_bw()

ggarrange(p1, p2, nrow = 2)
```



## Missing data


```{r parsinggeno, echo=FALSE, message=FALSE, warning=FALSE}
gt.geno$GT = unlist(lapply(strsplit(gt.geno[,2], ":"), `[[`, 1))
gt.geno$AD = unlist(lapply(strsplit(gt.geno[,2], ":"), `[[`, 7))
gt.geno$DP = unlist(lapply(strsplit(gt.geno[,2], ":"), `[[`, 6))
gt.geno$AD_REF = as.numeric(unlist(lapply(strsplit(gt.geno$AD, ","), `[[`, 1)))
gt.geno$AD_ALT = as.numeric(unlist(lapply(strsplit(gt.geno$AD, ","), `[[`, 2)))
gt.geno$PL = unlist(lapply(strsplit(gt.geno[,2], ":"), `[[`, 8))
gt.geno$PL_HOMREF = as.numeric(unlist(lapply(strsplit(gt.geno$PL, ","), `[[`, 1)))
gt.geno$PL_HET = as.numeric(unlist(lapply(strsplit(gt.geno$PL, ","), `[[`, 2)))
gt.geno$PL_HOMALT = as.numeric(unlist(lapply(strsplit(gt.geno$PL, ","), `[[`, 3)))
```


```{r missinggeno}
cat("Proportion of missing data (./.):", sum(gt.geno$GT == "./.")/nrow(gt.geno), "\n")

cat("Proportion of partially missing data (./?):", sum(gt.geno$GT %in% c("./0", "0/.", "1/.", "./1"))/nrow(gt.geno), "(should be 0 if everything is ok) \n")
```



## Heterozygosity


```{r HeGeno}
cat("Proportion of heterozygous SV (0/1 or 1/0):", sum(gt.geno$GT %in% c("1/0", "0/1"))/nrow(gt.geno), "\n")
```


We investigated the usual Variant Calling quality criterion:

* DP: Approximate read depth; some reads may have been filtered
* AD: Allelic depths for the ref and alt alleles in the order listed
* PL: Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification (lower score is better, lower score = most likely genotype)


## Supporting reads (DP:AD)



Approximate read depth (should be balanced between reference and variant if unbiased).


```{r SV_DP, echo = F, warning = F, fig.cap = "Genotype quality. Variant allele read depth (AD) as a function of the genotype."}
ggplot(gt.geno, aes(x = AD_REF, y = AD_ALT, group = GT, colour = GT)) +
  geom_point(alpha = 0.3) +
  xlab("Reference allele read depth (AD)") + ylab("Variant allele read depth (AD)") +
  theme_bw()
```


Variant allelic depth of the ALT allele (higher for heterozygotes is better).


```{r AD_GT, echo = F, warning = F}
gt.geno$GT = as.factor(gt.geno$GT)

p1 = ggplot(gt.geno, aes(x = GT, y = AD_ALT, group = GT, colour = GT)) +
  geom_boxplot() +
  ylab("Variant allelic depth (AD)") +
  theme_bw()

p2 = ggplot(gt.geno, aes(x = GT, y = AD_ALT, group = GT, colour = GT)) +
  geom_boxplot() +
  ylab("Variant allelic depth (AD)") +
  ylim(0, 100) +
  theme_bw()

ggarrange(p1, p2, ncol = 2)
```


## Phred-scale likelihoods for heterozygotes



Scores for 0/0 (homozygous ref), 0/1 (heterozygous), and 1/1 (homozygous alt) genotypes. For a phred-scaled likelihood of P, the raw likelihood of that genotype L = 10-P/10 (so the higher the number, the less likely it is that your sample is that genotype). A lower phred-scale likelihood is better. The sum of likelihoods is not necessarily 1. 



```{r PL, echo = F, warning = F}
df = data.frame(GT = c(rep("0/0", sum(gt.geno$GT %in% c("0/0"))),
                       rep("0/1", sum(gt.geno$GT %in% c("0/1"))),
                       rep("1/1", sum(gt.geno$GT %in% c("1/1")))),
                PL = c(gt.geno$PL_HOMREF[which(gt.geno$GT %in% c("0/0"))],
                       gt.geno$PL_HET[which(gt.geno$GT %in% c("0/1"))],
                       gt.geno$PL_HOMALT[which(gt.geno$GT %in% c("1/1"))]))
df$GT = as.factor(df$GT)


p1 = ggplot(df, aes(x = GT, y = PL, group = GT, colour = GT)) +
  geom_boxplot() +
  ylab("Phred-scaled genotype likelihood") +
  theme_bw()

p2 = ggplot(df, aes(x = GT, y = PL, group = GT, colour = GT)) +
  geom_boxplot() +
  ylab("Phred-scaled genotype likelihood") +
  ylim(0, 100) +
  theme_bw()

ggarrange(p1, p2, ncol = 2)
```


Now we plot the likelihood of the heterozygote for the three genotypes inferred. We expect a very low PL value for hetozygotes while PL values should be higher for HOM genotypes that are most likely calling errors. The difference in PL values between HET and HOM genotypes can be leveraged as a cutoff to filter low quality SV calls.



```{r PLHET, echo = F, warning = F}
df = data.frame(GT = c(rep("0/0", sum(gt.geno$GT %in% c("0/0"))),
                       rep("0/1", sum(gt.geno$GT %in% c("0/1"))),
                       rep("1/1", sum(gt.geno$GT %in% c("1/1")))),
                PL = c(gt.geno$PL_HET[which(gt.geno$GT %in% c("0/0"))],
                       gt.geno$PL_HET[which(gt.geno$GT %in% c("0/1"))],
                       gt.geno$PL_HET[which(gt.geno$GT %in% c("1/1"))]))
df$GT = as.factor(df$GT)


p1 = ggplot(df, aes(x = GT, y = PL, group = GT, colour = GT)) +
  geom_boxplot() +
  ylab("HET Phred-scaled genotype likelihood") +
  theme_bw()

p2 = ggplot(df, aes(x = GT, y = PL, group = GT, colour = GT)) +
  geom_boxplot() +
  ylab("HET Phred-scaled genotype likelihood") +
  ylim(0, 100) +
  theme_bw()

ggarrange(p1, p2, ncol = 2)
```





# Precision/recall for each tool and each type of SV


Estimate the calling rates for each tool:


* the TRUE POSITIVE rate is when a call is genotyped as HET by SVjedi-graph
* the FALSE POSITIVE rate is when a call is genotyped as HOM by SVjedi-graph
* the FALSE NEGATIVE rate is when a call is genotyped as missing (./.) while it has been called as HET for at least one other tool (we assume that a HET call is a true call, yet it an approximation and ground truth remains unknown). Note that because of this approximation we might be underestimating the false negative rate (the proportion of undetected variants) and overestimating the recall and F1 score. However, without ground truth, it is difficult to estimate the false negative rate, and our performance score is still valid to compare and weight the different tools.
* the TRUE NEGATIVE rate is not estimated, and is not used in precision/recall calculation




Quick reminder, the precision is TP / (TP + FP) and recall is TP / (TP +  FN). The F1 score is then F1 = 2 * ((precision * recall) / (precision + recall)).


```{r, echo = T}
# Build a matrix of calls per tool
# "minimap2_sniffles", "minimap2_svim", "minimap2_cutesv",
#                                    "minimap2_debreak", "ngmlr_sniffles", "ngmlr_svim",
#                                    "ngmlr_cutesv", "ngmlr_debreak"
genotype = fix.merged[,c(1:3, 7)]

gen = unlist(lapply(strsplit(gt.merged$minimap2_sniffles, ":"), `[[`, 1))
gen = ifelse(gen == "0/1", "HET", ifelse(gen == "./.", "MISSING", "HOM") )
genotype$minimap2_sniffles = gen

gen = unlist(lapply(strsplit(gt.merged$minimap2_svim, ":"), `[[`, 1))
gen = ifelse(gen == "0/1", "HET", ifelse(gen == "./.", "MISSING", "HOM") )
genotype$minimap2_svim = gen

gen = unlist(lapply(strsplit(gt.merged$minimap2_cutesv, ":"), `[[`, 1))
gen = ifelse(gen == "0/1", "HET", ifelse(gen == "./.", "MISSING", "HOM") )
genotype$minimap2_cutesv = gen

gen = unlist(lapply(strsplit(gt.merged$minimap2_debreak, ":"), `[[`, 1))
gen = ifelse(gen == "0/1", "HET", ifelse(gen == "./.", "MISSING", "HOM") )
genotype$minimap2_debreak = gen



gen = unlist(lapply(strsplit(gt.merged$ngmlr_sniffles, ":"), `[[`, 1))
gen = ifelse(gen == "0/1", "HET", ifelse(gen == "./.", "MISSING", "HOM") )
genotype$ngmlr_sniffles = gen

gen = unlist(lapply(strsplit(gt.merged$ngmlr_svim, ":"), `[[`, 1))
gen = ifelse(gen == "0/1", "HET", ifelse(gen == "./.", "MISSING", "HOM") )
genotype$ngmlr_svim = gen

gen = unlist(lapply(strsplit(gt.merged$ngmlr_cutesv, ":"), `[[`, 1))
gen = ifelse(gen == "0/1", "HET", ifelse(gen == "./.", "MISSING", "HOM") )
genotype$ngmlr_cutesv = gen

gen = unlist(lapply(strsplit(gt.merged$ngmlr_debreak, ":"), `[[`, 1))
gen = ifelse(gen == "0/1", "HET", ifelse(gen == "./.", "MISSING", "HOM") )
genotype$ngmlr_debreak = gen

# Estimate precision, recall and F1 score
performance = data.frame(tool = c("minimap2_sniffles", "minimap2_svim", "minimap2_cutesv",
                                   "minimap2_debreak", "ngmlr_sniffles", "ngmlr_svim",
                                   "ngmlr_cutesv", "ngmlr_debreak"),
                         n_calls = NA,
                         true_positive = NA,
                         false_positive = NA,
                         false_negative = NA)

n_calls = nrow(genotype)

for (i in 1:nrow(performance)) {
  performance$n_calls[i] = sum(genotype[,performance$tool[i]] != "MISSING")
  performance$true_positive[i] = sum(genotype[,performance$tool[i]] == "HET")
  performance$false_positive[i] = sum(genotype[,performance$tool[i]] == "HOM")
  het_in_one_tool = apply(genotype, 1, function(x) {"HET" %in% x})
  performance$false_negative[i] = sum(genotype[,performance$tool[i]] == "MISSING" & het_in_one_tool)
}

performance$precision = performance$true_positive / (performance$true_positive + performance$false_positive)
performance$recall = performance$true_positive / (performance$true_positive + performance$false_negative)

performance$f1 = 2 * ((performance$precision * performance$recall) / (performance$precision + performance$recall))
```


<!-- Now estimate the performance for each type of SV (DEL, INS, DUP, INV). -->


```{r, echo = F}
performance$sv_type = "all"


for (type in c("DEL", "INS", "DUP", "INV")) {
  # Estimate precision, recall and F1 score
  performance_per_type = data.frame(tool = c("minimap2_sniffles", "minimap2_svim", "minimap2_cutesv",
                                     "minimap2_debreak", "ngmlr_sniffles", "ngmlr_svim",
                                     "ngmlr_cutesv", "ngmlr_debreak"),
                           n_calls = NA,
                           true_positive = NA,
                           false_positive = NA,
                           false_negative = NA)
  
  genotype_subset = genotype[genotype$sv_type == type,]
  # n_calls = nrow(genotype_subset)
  
  for (i in 1:nrow(performance_per_type)) {
    performance_per_type$n_calls[i] = sum(genotype_subset[,performance_per_type$tool[i]] != "MISSING")
    performance_per_type$true_positive[i] = sum(genotype_subset[,performance_per_type$tool[i]] == "HET")
    performance_per_type$false_positive[i] = sum(genotype_subset[,performance_per_type$tool[i]] == "HOM")
    het_in_one_tool = apply(genotype_subset, 1, function(x) {"HET" %in% x})
    performance_per_type$false_negative[i] = sum(genotype_subset[,performance_per_type$tool[i]] == "MISSING" & het_in_one_tool)
  }
  
  performance_per_type$precision = performance_per_type$true_positive / (performance_per_type$true_positive + performance_per_type$false_positive)
  performance_per_type$recall = performance_per_type$true_positive / (performance_per_type$true_positive + performance_per_type$false_negative)
  
  performance_per_type$f1 = 2 * ((performance_per_type$precision * performance_per_type$recall) / (performance_per_type$precision + performance_per_type$recall))
  
  performance_per_type$sv_type = type
  
  performance = rbind(performance, performance_per_type)
}
```



```{r, echo = T, tab.cap = "Performance of the different SV calling tools for each SV type and all SV pooled."}
kable(performance, digits = 3, align = "c")
```





```{r, echo = F}
# Save the performance output
write.table(performance, paste0("../../", wdir, "/performance/", genome, "_performance.tsv"),
            quote = F, sep = "\t", row.names = F, col.names = T)
```





## Likelihood Ratio Test between the best likelihood and the HET likelihood


An alternative measure of confidence for each call.





# SV and genomic landscapes



## Quality and support along the genome

Quality scores and allele support are expected to homogeneous along the genome.


Quality Score.

```{r QClandscape, echo = F, warning = F}
ggplot(fix.merged, aes(x = as.numeric(POS)/10^6, y = as.numeric(QUAL))) +
  geom_point(alpha = 0.5) +
  geom_smooth(method = "gam", se = T) +
  xlab("Position (Mb)") + ylab("SV Quality score") +
  facet_wrap(~ CHROM, scales = "free_y", ncol = 3) +
  theme_bw()
```


Support for variant allele.


```{r suppportlandscape, echo = F, warning = F}
fix.merged$sv_support = unlist(lapply(fix.merged$supp_vec, function(x) sum(as.numeric(strsplit(x, "")[[1]]))))

ggplot(fix.merged, aes(x = as.numeric(POS)/10^6, y = as.numeric(sv_support))) +
  geom_point(alpha = 0.5) +
  geom_smooth(method = "gam", se = T) +
  xlab("Position (Mb)") + ylab("SV support for allele variant") +
  facet_wrap(~ CHROM, scales = "free_y", ncol = 3) +
  scale_y_continuous(breaks = c(1:8), labels = as.character(1:8)) +
  theme_bw()
```






## The distribution of SVs along the genome


```{r SVlandscape, echo = F, warning = F}
ggplot(fix.merged, aes(x = as.numeric(POS)/10^6)) +
  geom_histogram(bins = 1000) +
  xlab("Position (Mb)") + ylab("SV count") +
  facet_wrap(~ CHROM, ncol = 3)
```




